hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,319 @@
+			final MerkleAccount account, final long accountNum) {
+		if (changes.containsKey(AccountProperty.BALANCE)) {
+			final long newBalance = (long) changes.get(AccountProperty.BALANCE);
+			if (account != null && (accountNum == STAKING_FUNDING_ACCOUNT_NUMBER)) {
+				rewardBalanceChanged = true;
+				newRewardBalance = newBalance;
+			}
+		}
+	}
+
+	private void trackRewardsPaidByFunding(
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		final var rewardsPaid = rewardCalculator.rewardsPaidInThisTxn();
+		final var rewardAccountI = findOrAdd(800L, pendingChanges);
+		manager.updateBalance(-rewardsPaid, rewardAccountI, pendingChanges);
+	}
+
+	private void updateNodeStakes(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
+			final var account = pendingChanges.entity(i);
+			final var changes = pendingChanges.changes(i);
+
+			final var curNodeId = (account != null) ? account.getStakedId() : 0L;
+			final var newNodeId = manager.getNodeStakeeNum(changes);
+			if (curNodeId < 0 && curNodeId != newNodeId) {
+				// Node stakee has been replaced, withdraw stakeRewarded or stakeNotRewarded from ex-stakee based on
+				// isDeclineReward option
+				manager.withdrawStake(
+						Math.abs(curNodeId),
+						account.getBalance() + account.getStakedToMe(),
+						manager.finalDeclineRewardGiven(account, changes));
+			}
+			if (newNodeId < 0) {
+				// Award updated stake to new node stakee to the fields stakeRewarded or stakeNotRewarded from
+				// ex-stakee based on isDeclineReward option
+				manager.awardStake(
+						Math.abs(newNodeId),
+						finalBalanceGiven(account, changes) + manager.finalStakedToMeGiven(account, changes),
+						manager.finalDeclineRewardGiven(account, changes));
+			}
+		}
+	}
+
+	int updateStakedToMeSideEffects(
+			final int num,
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges,
+			final Set<Long> hasBeenRewarded,
+			final long latestEligibleStart) {
+		int changesSize = pendingChanges.size();
+		final var account = pendingChanges.entity(num);
+		final var changes = pendingChanges.changes(num);
+
+		final var curStakeeNum = (account != null) ? account.getStakedId() : 0L;
+		final var newStakeeNum = manager.getAccountStakeeNum(changes);
+
+		if (curStakeeNum > 0 && curStakeeNum != newStakeeNum) {
+			// Stakee has been replaced, withdraw initial balance from ex-stakee
+			final var exStakeeI = findOrAdd(curStakeeNum, pendingChanges);
+			manager.updateStakedToMe(exStakeeI, -account.getBalance(), pendingChanges);
+			if (exStakeeI == changesSize) {
+				changesSize++;
+			} else if (!hasBeenRewarded.contains(curStakeeNum)) {
+				payRewardIfRewardable(pendingChanges, exStakeeI, hasBeenRewarded, latestEligibleStart);
+				// do we need this ?? since stakedToMe is only for account stakes
```

Comment:
I think we need it because, in theory, the changes at `exStakeeI` may have returned `false` from `hasStakeFieldChanges()`.